src/cla/tcp.rs

```diff
@@ -0,0 +1,511 @@
+                let (packet_sender_send, packet_receiver_send) = channel::<TcpClPacket>(50);
+                let (session_sender, session_receiver) = channel::<(SocketAddr, ByteBuffer)>(50);
+                let (tcp_receiver, tcp_sender) = self.stream.into_split();
+                let receiver_task = TcpReceiver {
+                    stream_reader: tcp_receiver,
+                    session_channel: packet_sender_rcv,
+                    timeout: remote_parameters.keepalive,
+                };
+                let sender_task = TcpSender {
+                    stream_writer: tcp_sender,
+                    session_channel: packet_receiver_send,
+                    timeout: local_parameters.keepalive,
+                };
+                let session_task = TcpSession {
+                    sender_task: packet_sender_send,
+                    receiver_task: packet_receiver_rcv,
+                    bundle_queue: session_receiver,
+                    data_local: local_parameters,
+                    data_remote: remote_parameters,
+                    last_tid: 0,
+                };
+                tokio::spawn(receiver_task.run());
+                tokio::spawn(sender_task.run());
+                tokio::spawn(session_task.run());
+                debug!("Spawned TCP session");
+                sessions.insert(self.addr, session_sender);
+                debug!("Started TCP session");
+            }
+            Err(err) => error!("Failed to negotiate session: {}", err),
+        }
+    }
+}
+
+pub struct Listener {
+    tcp_listener: TcpListener,
+    sessions: Arc<Mutex<HashMap<SocketAddr, Sender<(SocketAddr, ByteBuffer)>>>>,
+}
+
+impl Listener {
+    async fn run(self) {
+        loop {
+            match self.tcp_listener.accept().await {
+                Ok((stream, addr)) => {
+                    info!("Incoming connection from: {:?}", addr);
+                    let connection = Connection { stream, addr };
+                    connection.connect(&mut self.sessions.lock().await).await;
+                }
+                Err(e) => {
+                    error!("Couldn't get client: {:?}", e)
+                }
+            }
+        }
+    }
+}
+
+impl TcpConvergenceLayer {
+    pub fn new(port: Option<u16>) -> TcpConvergenceLayer {
+        TcpConvergenceLayer {
+            local_port: port.unwrap_or(4556),
+            sessions: Arc::new(Mutex::new(HashMap::new())),
+            listener: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    async fn send_bundles(
+        &self,
+        addr: &SocketAddr,
+        ready: &[ByteBuffer],
+    ) -> Result<bool, SendError<(SocketAddr, ByteBuffer)>> {
+        if let Some(sender) = self.sessions.lock().await.get(&addr) {
+            debug!("Using existing connection for {}", addr);
+            for bundle in ready {
+                debug!("Sending bundle {:?}", bundle);
+                sender.send((addr.clone(), bundle.clone())).await?
```

Comment:
can't the bundle cloning be eliminated here? a reference to the bytebuffer should be enough as not much processing is done but "just" written to the remote peer